File: backend/models/media_list_entry.go
package models

import pb "goodanime-backend/proto"

// ToProto converts MediaListEntry to its protobuf representation.
func (e *MediaListEntry) ToProto() *pb.MediaListEntry {
	if e == nil {
		return nil
	}

	return &pb.MediaListEntry{
		Id:          e.ID,
		Status:      e.Status.ToProto(),
		Progress:    e.Progress,
		Score:       e.Score,
		Repeat:      e.Repeat,
		StartedAt:   e.StartedAt.ToProto(),
		CompletedAt: e.CompletedAt.ToProto(),
	}
}

// MediaListEntry represents a media list entry
type MediaListEntry struct {
	ID          int32            `json:"id"`
	Status      *MediaListStatus `json:"status"`
	Progress    int32            `json:"progress"`
	Score       float64          `json:"score"`
	Repeat      int32            `json:"repeat"`
	StartedAt   FuzzyDate        `json:"startedAt"`
	CompletedAt FuzzyDate        `json:"completedAt"`
}

// SaveMediaListEntryDTO represents the DTO for saving a list entry
type SaveMediaListEntryDTO struct {
	SaveMediaListEntry MediaListEntry `json:"SaveMediaListEntry"`
}

// DeleteMediaListEntry represents the response when deleting a list entry
type DeleteMediaListEntry struct {
	Deleted bool `json:"deleted"`
}

// DeleteMediaListEntryDTO represents the DTO for deleting a list entry
type DeleteMediaListEntryDTO struct {
	DeleteMediaListEntry DeleteMediaListEntry `json:"DeleteMediaListEntry"`
}

File: lib/models/media_list_entry.dart
import 'package:protobuf/protobuf.dart';
import '../proto/media_list_entry.pb.dart' as pb;
import 'common.dart';

/// Represents a media list entry
class MediaListEntry {
  /// Entry ID
  final int id;

  /// Watch status
  final MediaListStatus? status;

  /// Number of episodes watched
  final int progress;

  /// User score
  final double score;

  /// Number of rewatches
  final int repeat;

  /// Number of volumes read
  final int progressVolumes;

  /// Date the user started watching
  final FuzzyDate? startedAt;

  /// Date the user completed watching
  final FuzzyDate? completedAt;

  /// Creates a media list entry
  const MediaListEntry({
    required this.id,
    this.status,
    required this.progress,
    required this.score,
    required this.repeat,
    required this.progressVolumes,
    this.startedAt,
    this.completedAt,
  });

  /// Creates a media list entry from a protobuf object
  factory MediaListEntry.fromProto(pb.MediaListEntry pbObj) {
    return MediaListEntry(
      id: pbObj.id,
      status: pbObj.hasStatus()
          ? MediaListStatus.fromProto(pbObj.status)
          : null,
      progress: pbObj.progress,
      score: pbObj.score,
      repeat: pbObj.repeat,
      progressVolumes: pbObj.progressVolumes,
      startedAt: pbObj.hasStartedAt()
          ? FuzzyDate.fromProto(pbObj.startedAt)
          : null,
      completedAt: pbObj.hasCompletedAt()
          ? FuzzyDate.fromProto(pbObj.completedAt)
          : null,
    );
  }

  /// Converts the media list entry to a protobuf object
  GeneratedMessage toProto() {
    final pbObj = pb.MediaListEntry(
      id: id,
      progress: progress,
      score: score,
      repeat: repeat,
      progressVolumes: progressVolumes,
    );
    if (status != null) pbObj.status = status!.toProto();
    if (startedAt != null) pbObj.startedAt = startedAt!.toProto();
    if (completedAt != null) pbObj.completedAt = completedAt!.toProto();
    return pbObj;
  }

  /// Creates a copy of this object with the given fields replaced
  MediaListEntry copyWith({
    int? id,
    MediaListStatus? status,
    int? progress,
    double? score,
    int? repeat,
    int? progressVolumes,
    FuzzyDate? startedAt,
    FuzzyDate? completedAt,
  }) {
    return MediaListEntry(
      id: id ?? this.id,
      status: status ?? this.status,
      progress: progress ?? this.progress,
      score: score ?? this.score,
      repeat: repeat ?? this.repeat,
      progressVolumes: progressVolumes ?? this.progressVolumes,
      startedAt: startedAt ?? this.startedAt,
      completedAt: completedAt ?? this.completedAt,
    );
  }
}
